### PR TITLE
Implement FracBitsQuantizationBuilder and Controller

### PR DESCRIPTION
**Pull request by vinnamkim**
This should be merged after #1231.

### Changes
 - Implement Builder and Controller
 - Add and test ModelSizeCompressionLoss

### Reason for changes
 - To implement the initial version of FracBitsQuantizationBuilder, Controller, and Loss.

### Related tickets
87888, 87889

### Tests
Related unit tests added

### CODE REVIEW

**Review by alexsu52 — CHANGES_REQUESTED** [review]
@vinnamkim, thanks for your contribution! I have major comments/questions about this PR and https://github.com/openvinotoolkit/nncf/pull/1231. Could you share what results you have achieved with the FracBits quantization algorithm? What user cases are you going to cover by this algorithm? What is the benefit?

In accordance with the contribution guide for NNCF. The PRs with fully implemented features are only approved. All research must be done on forks.

cc' @AlexKoff88

**Comment by vinnamkim** [comment]
> @vinnamkim, thanks for your contribution! I have major comments/questions about this PR and #1231. Could you share what results you have achieved with the FracBits quantization algorithm? What user cases are you going to cover by this algorithm? What is the benefit?

First of all, as far as I know this item is agreed with Lee, Wonju and Kozlov, Alexander on seeing results after quick implementation. I'm working now to obtain the experiment results on MobileNetV2, and etc... Basically, it can extend user choice for QAT algorithms. Currently, we has been provided only two algorithm: HaWQ and AutoQ which has an additional exploration stage aside QAT. Fracbits provides fully differential training for bit-widths, which is completely different from HaWQ and AutoQ. 

Therefore, this algorithm's benefit is exactly same for HaWQ and AutoQ. On top of that, I think the mixed precision will be more important in the future. This is because the model size of the vision models that NNCF has dealt with so far was so small if we compared to Transformers (BERT-Large ~1.4G). However, Transformers becomes the major player in AI scene and I believe there is demand from users to compress their deployment model size. We all know that commercial hardware acceleration for NN is supported up to 8bit, and OV runtime doesn't allow to save model weights on the mixed precision. But, I think it's easy to implement at the software level.

> In accordance with the contribution guide for NNCF. The PRs with fully implemented features are only approved. All research must be done on forks.

I know this was against the rules. How about make a branching strategy to use a feature branch. We can update and review code in a fast cycle on the feature branch. If all work ends in that feature branch, we can try to merge that feature branch into the develop branch. This is because trying to merge large PRs at once should be avoided in terms of software development. What do you think?

**Comment by alexsu52** [comment]
> merge that feature branch into the develop branch. This is because trying to merge large PRs at once should be avoided in terms of software development. What do you think?

Hi, @vinnamkim. Thank you for your response! I have no concerns about researching in the mixed precision area, in particular, implementing FracBits quantization algorithm. My concern is that the PR must meet the contribution guide and be justified in order to be accepted for merging with develop.

The fork strategy is a default stretage for NNCF and uses in many repositories under the OpenVINO umbrella. If you do not like the fork strategy, I think you can create a branch in the NNCF repository to implement the research feature.

File: nncf/experimental/torch/fracbits/loss.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                return (module.weight.shape.numel() * num_bits).sum()
+            nncf_logger.warning("module={module} is not supported by ModelSizeCompressionLoss. Skip it.")
+            return 0.
+
+        return sum([_get_module_size(pair.module, pair.quantizer.frac_num_bits) for pair in self._w_q_pairs.values()])
+
+    @torch.no_grad()
+    def get_state(self) -> Dict[str, Number]:
+        states = {
+            "compression_rate": self._init_model_size / (self._get_model_size() + 1e-6).item()
```
Comment (wonjuleee) [inline nncf/experimental/torch/fracbits/loss.py:83]: Here also.

File: tests/torch/experimental/fracbits/test_builder.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+from nncf.experimental.torch.fracbits.builder import FracBitsQuantizationBuilder
+from nncf.experimental.torch.fracbits.quantizer import FracBitsSymmetricQuantizer
+
+
+def test_create_builder(config):
+    builder = create_compression_algorithm_builder(config)
+    assert isinstance(builder, FracBitsQuantizationBuilder)
+
+
+def test_can_load_quant_algo__with_defaults(config, conv_model):
```
Comment (wonjuleee) [inline tests/torch/experimental/fracbits/test_builder.py:34]: Small minor typo `algo__with`

File: tests/torch/experimental/fracbits/test_builder.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    # Check whether bit_width gradient is not None
+    loss_value.backward()
+    for qinfo in compression_ctrl.weight_quantizers.values():
+        q = qinfo.quantizer_module_ref
+        assert q._num_bits.grad.data is not None
+
+
+def test_e2e_quant_loss(config, conv_model_with_input_output):
+    conv_model, x, y = conv_model_with_input_output
+    criterion = torch.nn.MSELoss()
```
Comment (wonjuleee) [inline tests/torch/experimental/fracbits/test_builder.py:75]: Is this general to use MSE loss for target and L1 loss for compression?

File: tests/torch/experimental/fracbits/test_builder.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    quant_model, compression_ctrl = create_compressed_model_and_algo_for_test(conv_model, config)
+
+    optimizer = torch.optim.SGD(quant_model.parameters(), lr=1e-1)
+
+    for i in range(500):
```
Comment (wonjuleee) [inline tests/torch/experimental/fracbits/test_builder.py:81]: Is this okay to run 500 iterations for unittest?

**Comment by wonjuleee** [comment]
> > merge that feature branch into the develop branch. This is because trying to merge large PRs at once should be avoided in terms of software development. What do you think?
> 
> Hi, @vinnamkim. Thank you for your response! I have no concerns about researching in the mixed precision area, in particular, implementing FracBits quantization algorithm. My concern is that the PR must meet the contribution guide and be justified in order to be accepted for merging with develop.
> 
> The fork strategy is a default strategy for NNCF and uses in many repositories under the OpenVINO umbrella. If you do not like the fork strategy, I think you can create a branch in the NNCF repository to implement the research feature.

Hi @alexsu52, FracBits feature should be merged to the develop branch in a viewpoint of "implement a recently published compression algorithm" in the contribution guide. If only the fully implemented feature including example and recipe can be merged into the develop, @vinnamkim, could you work on your forked branch and make PRs to implement all related to FracBits by separating PRs (small PR must be preferred) and merge this branch to the develop. @alexsu52, is this okay to you?

**Comment by vinnamkim** [comment]
I changed this PR's target branch to `feature/fracbits` and made a revert PR https://github.com/openvinotoolkit/nncf/pull/1237. We can continue this review process by uploading PR to `feature/fracbits` branch. So, when all works are done, we can make an another PR to merge `feature/fracbits` into `develop` branch.

File: tests/torch/experimental/fracbits/test_builder.py (hunk 3 above)
Comment (vinnamkim) [reply tests/torch/experimental/fracbits/test_builder.py:75]: MSELoss is just for tests. FracBits uses L1 loss in their paper.

File: tests/torch/experimental/fracbits/test_builder.py (hunk 4 above)
Comment (vinnamkim) [reply tests/torch/experimental/fracbits/test_builder.py:81]: Yes. Because we fixed the random seed, it will reach the target compression rate every time.

File: nncf/experimental/torch/fracbits/loss.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                parent_name = ".".join(name.split(".")[:-2])
+                parent_module = self._model.get_submodule(parent_name)
+
+                self._w_q_pairs[parent_name] = ModuleQuantizerPair(parent_module, module.op)
+
+        with torch.no_grad():
+            self._init_model_size = self._get_model_size()
+
+    def calculate(self) -> torch.Tensor:
+        cur_comp_rate = self._init_model_size / (self._get_model_size() + 1e-6)
```
[the remark this answers is not in the document]
Comment (vinnamkim) [reply nncf/experimental/torch/fracbits/loss.py]: Fixed.

File: nncf/experimental/torch/fracbits/loss.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        if criteria == "L2":
+            return nn.MSELoss()
+        raise RuntimeError(f"Unknown criteria = {criteria}.")
+
+    def _get_model_size(self) -> Union[torch.Tensor, Number]:
+        def _get_module_size(module: nn.Module, num_bits: Union[int, torch.Tensor]) -> Union[torch.Tensor, Number]:
+            if isinstance(module, nn.modules.conv._ConvNd):
+                return (module.weight.shape.numel() * num_bits).sum()
+            if isinstance(module, nn.Linear):
+                return (module.weight.shape.numel() * num_bits).sum()
```
[the remark this answers is not in the document]
Comment (vinnamkim) [reply nncf/experimental/torch/fracbits/loss.py:73]: Fixed.

File: nncf/experimental/torch/fracbits/loss.py (hunk 1 above)
Comment (vinnamkim) [reply nncf/experimental/torch/fracbits/loss.py]: Fixed.

File: tests/torch/experimental/fracbits/test_builder.py (hunk 2 above)
Comment (vinnamkim) [reply tests/torch/experimental/fracbits/test_builder.py]: Fixed.

**Comment by alexsu52** [comment]
> 

Hi, @wonjuleee. Thanks for your comment. I totally agree with you.

**Review by wonjuleee — APPROVED** [review]
LGTM.
@vinnamkim, will you make a new PR for FracBits algorithm implementation?
Is this okay to merge this first?

**Comment by vinnamkim** [comment]
> LGTM. @vinnamkim, will you make a new PR for FracBits algorithm implementation? Is this okay to merge this first?

Yes, some remaining works may be uploaded as PRs to [feature/fracbits](https://github.com/openvinotoolkit/nncf/tree/feature/fracbits) branch. And if all works is done, we can try merge [feature/fracbits](https://github.com/openvinotoolkit/nncf/tree/feature/fracbits) to develop branch.

**Comment by vinnamkim** [comment]
> > merge that feature branch into the develop branch. This is because trying to merge large PRs at once should be avoided in terms of software development. What do you think?
> 
> Hi, @vinnamkim. Thank you for your response! I have no concerns about researching in the mixed precision area, in particular, implementing FracBits quantization algorithm. My concern is that the PR must meet the contribution guide and be justified in order to be accepted for merging with develop.
> 
> The fork strategy is a default strategy for NNCF and uses in many repositories under the OpenVINO umbrella. If you do not like the fork strategy, I think you can create a branch in the NNCF repository to implement the research feature.

@alexsu52 
I see that your major claim for the change request is that this PR targets to `develop` branch. Because the target branch for this PR is changed, I think that we can merge it now.
